File: SPCServer/springboot/src/main/java/com/lcaohoanq/shoppe/dtos/responses/base/ApiResponse.java
package com.lcaohoanq.shoppe.dtos.responses.base;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

@NoArgsConstructor
@AllArgsConstructor
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ApiResponse<T> {
    @JsonProperty("status_code")
    protected Integer statusCode;

    @JsonProperty("message")
    protected String message;

    @JsonProperty("reason")
    protected String reason;

    @JsonProperty("is_success")
    protected Boolean isSuccess;

    @JsonProperty("data")
    protected T data;

    public static <T> ApiResponseBuilder<T> builder() {
        return new ApiResponseBuilder<>();
    }

    public static class ApiResponseBuilder<T> {
        private Integer statusCode;
        private String message;
        private String reason;
        private Boolean isSuccess;
        private T data;

        ApiResponseBuilder() {
        }

        public ApiResponseBuilder<T> statusCode(Integer statusCode) {
            this.statusCode = statusCode;
            return this;
        }

        public ApiResponseBuilder<T> message(String message) {
            this.message = message;
            return this;
        }

        public ApiResponseBuilder<T> reason(String reason) {
            this.reason = reason;
            return this;
        }

        public ApiResponseBuilder<T> isSuccess(Boolean isSuccess) {
            this.isSuccess = isSuccess;
            return this;
        }

        public ApiResponseBuilder<T> data(T data) {
            this.data = data;
            return this;
        }

        public ApiResponse<T> build() {
            return new ApiResponse<>(statusCode, message, reason, isSuccess, data);
        }
    }
}

File: product-service/src/main/kotlin/com/lcaohoanq/productservice/api/ApiResponse.kt
package com.lcaohoanq.productservice.api

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import io.swagger.v3.oas.annotations.media.Schema

@JsonPropertyOrder("message", "data", "status_code", "is_success", "reason")
@JsonInclude(JsonInclude.Include.NON_NULL)
@Schema(description = "API Response")
data class ApiResponse<T>(
    @JsonProperty("status_code")
    var statusCode: Int? = null,

    @JsonProperty("message")
    @Schema(description = "Response message", example = "Login successfully")
    var message: String? = null,

    @JsonProperty("reason")
    @Schema(description = "Reason of response", example = "User not found")
    var reason: String? = null,

    @JsonProperty("is_success")
    @Schema(description = "Response status", example = "true")
    var isSuccess: Boolean? = null,

    @JsonProperty("data")
    @Schema(description = "Response data")
    var data: T? = null
) {

    companion object {
        fun <T> builder() = Builder<T>()
    }

    class Builder<T> {
        private var statusCode: Int? = null
        private var message: String? = null
        private var reason: String? = null
        private var isSuccess: Boolean? = null
        private var data: T? = null

        fun statusCode(statusCode: Int?) = apply { this.statusCode = statusCode }
        fun message(message: String?) = apply { this.message = message }
        fun reason(reason: String?) = apply { this.reason = reason }
        fun isSuccess(isSuccess: Boolean?) = apply { this.isSuccess = isSuccess }
        fun data(data: T?) = apply { this.data = data }

        fun build() = ApiResponse(
            statusCode = statusCode,
            message = message,
            reason = reason,
            isSuccess = isSuccess,
            data = data
        )
    }
}
